File: app/src/main/java/com/example/home/db/Converter.java
package com.example.home.db;

import androidx.room.TypeConverter;

import com.example.home.model.HighlightResult;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.List;

public class Converter {

    @TypeConverter
    public static String fromObject(Object object) {
        Gson gson = new Gson();
        return gson.toJson(object);
    }

    @TypeConverter
    public static Object toObject(String object){
        return new Gson().fromJson(object,Object.class);
    }

    @TypeConverter
    public static String fromHighlightResult(HighlightResult result){
        return new Gson().toJson(result);
    }

    @TypeConverter
    public static HighlightResult toHighlightResult(String object){
        return new Gson().fromJson(object,HighlightResult.class);
    }

    @TypeConverter
    public static List<String> toList(String value) {
        Type listType = new TypeToken<List<String>>() {}.getType();
        return new Gson().fromJson(value, listType);
    }
    @TypeConverter
    public static String fromList(List<String> list) {
        Gson gson = new Gson();
        String json = gson.toJson(list);
        return json;
    }

}

File: app/src/main/java/com/example/assignment13_mvvm_kotlin/converter/Converter.kt
package com.example.assignment13_mvvm_kotlin.converter

import androidx.room.TypeConverter
import com.example.assignment13_mvvm_kotlin.model.HighlightResult
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import java.lang.reflect.Type


class Converter {

    @TypeConverter
    fun fromObject(obj: Any?): String {
        val gson = Gson()
        return gson.toJson(obj)
    }

    @TypeConverter
    fun toObject(obj: String?): Any? {
        return Gson().fromJson(obj, Any::class.java)
    }

    @TypeConverter
    fun fromHighlightResult(result: HighlightResult?): String? {
        return Gson().toJson(result)
    }

    @TypeConverter
    fun toHighlightResult(obj: String?): HighlightResult? {
        return Gson().fromJson(obj, HighlightResult::class.java)
    }

    @TypeConverter
    fun toList(value: String?): List<String?>? {
        val listType: Type = object : TypeToken<List<String?>?>() {}.type
        return Gson().fromJson(value, listType)
    }

    @TypeConverter
    fun fromList(list: List<String?>?): String? {
        val gson = Gson()
        return gson.toJson(list)
    }
}